openpype/modules/kitsu/utils/update_op_with_zou.py

```diff
@@ -0,0 +1,372 @@
+
+    Args:
+        dbcon (AvalonMongoDB): Connection to DB.
+        project_id (str): Project zou ID
+    """
+    project = gazu.project.get_project(project_id)
+    project_name = project["name"]
+    dbcon.Session["AVALON_PROJECT"] = project_name
+
+
+def update_op_assets(
+    project_col: Collection,
+    entities_list: List[dict],
+    asset_doc_ids: Dict[str, dict],
+) -> List[Dict[str, dict]]:
+    """Update OpenPype assets.
+    Set 'data' and 'parent' fields.
+
+    Args:
+        project_col (Collection): Mongo project collection to sync
+        entities_list (List[dict]): List of zou entities to update
+        asset_doc_ids (Dict[str, dict]): Dicts of [{zou_id: asset_doc}, ...]
+
+    Returns:
+        List[Dict[str, dict]]: List of (doc_id, update_dict) tuples
+    """
+    project_name = project_col.name
+
+    assets_with_update = []
+    for item in entities_list:
+        # Update asset
+        item_doc = asset_doc_ids[item["id"]]
+        item_data = deepcopy(item_doc["data"])
+        item_data.update(item.get("data") or {})
+        item_data["zou"] = item
+
+        # == Asset settings ==
+        # Frame in, fallback on 0
+        frame_in = int(item_data.get("frame_in") or 0)
+        item_data["frameStart"] = frame_in
+        # Frame out, fallback on frame_in + duration
+        frames_duration = int(item.get("nb_frames") or 1)
+        frame_out = (
+            item_data["frame_out"]
+            if item_data.get("frame_out")
+            else frame_in + frames_duration
+        )
+        item_data["frameEnd"] = int(frame_out)
+        # Fps, fallback to project's value when entity fps is deleted
+        if not item_data.get("fps") and item_doc["data"].get("fps"):
+            project_doc = project_col.find_one({"type": "project"})
+            item_data["fps"] = project_doc["data"]["fps"]
+
+        # Tasks
+        tasks_list = []
+        item_type = item["type"]
+        if item_type == "Asset":
+            tasks_list = all_tasks_for_asset(item)
+        elif item_type == "Shot":
+            tasks_list = all_tasks_for_shot(item)
+            # TODO frame in and out
+        item_data["tasks"] = {
+            t["task_type_name"]: {"type": t["task_type_name"]}
+            for t in tasks_list
+        }
+
+        # Get zou parent id for correct hierarchy
+        # Use parent substitutes if existing
+        substitute_parent_item = (
+            item_data["parent_substitutes"][0]
+            if item_data.get("parent_substitutes")
+            else None
+        )
+        if substitute_parent_item:
+            parent_zou_id = substitute_parent_item["parent_id"]
+        else:
+            parent_zou_id = (
+                item.get("parent_id")
+                or item.get("episode_id")
+                or item.get("source_id")
+            )  # TODO check consistency
+
+        # Substitute Episode and Sequence by Shot
+        project_module_settings = get_project_settings(project_name)["kitsu"]
```

Comment:
I would recommend to not get project settings in each loop. The settings probably won't change but resolving project settings can slow the sync down a lot.

Reply:
Moved out of loop